### ISSUE TITLE
Cached files (previews, avatars, theming) should also be placed in object store

### ISSUE THREAD

**Issue by MorrisJobke** [nextcloud/server#4714]
* setup a object store instance
* upload some files and create some users

## Expected

`data/` should only contain `index.html`, `.htaccess`, `.ocdata` and `nextcloud.log`.

## Actual

A lot more files are in there, that are also in the file cache causing issues in a multi web server issue. Files for the App Store, avatars, cached CSS/JS files, theming files, previews.

<details>

```
./appdata_ocabcdef
./appdata_ocabcdef/appstore
./appdata_ocabcdef/appstore/apps.json
./appdata_ocabcdef/preview
./appdata_ghijkl
./appdata_ghijkl/avatar
./appdata_ghijkl/avatar/Quo0yae0
./appdata_ghijkl/avatar/user000
./appdata_ghijkl/avatar/user004
./appdata_ghijkl/avatar/user005
./appdata_ghijkl/avatar/user006
./appdata_ghijkl/avatar/user007
./appdata_ghijkl/avatar/user008
./appdata_ghijkl/avatar/user009
./appdata_ghijkl/avatar/user010
...
./appdata_ghijkl/avatar/user411
./appdata_ghijkl/css
./appdata_ghijkl/css/core
./appdata_ghijkl/css/core/header.css
./appdata_ghijkl/css/core/header.css.deps
./appdata_ghijkl/css/core/header.css.gzip
./appdata_ghijkl/css/core/server.css
./appdata_ghijkl/css/core/server.css.deps
./appdata_ghijkl/css/core/server.css.gzip
./appdata_ghijkl/css/core/share.css
./appdata_ghijkl/css/core/share.css.deps
./appdata_ghijkl/css/core/share.css.gzip
./appdata_ghijkl/css/core/styles.css
./appdata_ghijkl/css/core/styles.css.deps
./appdata_ghijkl/css/core/styles.css.gzip
./appdata_ghijkl/css/core/systemtags.css
./appdata_ghijkl/css/core/systemtags.css.deps
./appdata_ghijkl/css/core/systemtags.css.gzip
./appdata_ghijkl/css/files
./appdata_ghijkl/css/files/merged.css
./appdata_ghijkl/css/files/merged.css.deps
./appdata_ghijkl/css/files/merged.css.gzip
./appdata_ghijkl/css/files_sharing
./appdata_ghijkl/css/files_sharing/mergedAdditionalStyles.css
./appdata_ghijkl/css/files_sharing/mergedAdditionalStyles.css.deps
./appdata_ghijkl/css/files_sharing/mergedAdditionalStyles.css.gzip
./appdata_ghijkl/css/files_texteditor
./appdata_ghijkl/css/files_texteditor/merged.css
./appdata_ghijkl/css/files_texteditor/merged.css.deps
./appdata_ghijkl/css/files_texteditor/merged.css.gzip
./appdata_ghijkl/css/theming
./appdata_ghijkl/css/theming/theming.css
./appdata_ghijkl/css/theming/theming.css.deps
./appdata_ghijkl/css/theming/theming.css.gzip
./appdata_ghijkl/js
./appdata_ghijkl/js/comments
./appdata_ghijkl/js/comments/merged.js
./appdata_ghijkl/js/comments/merged.js.deps
./appdata_ghijkl/js/comments/merged.js.gzip
./appdata_ghijkl/js/core
./appdata_ghijkl/js/core/merged.js
./appdata_ghijkl/js/core/merged.js.deps
./appdata_ghijkl/js/core/merged.js.gzip
./appdata_ghijkl/js/core/merged-login.js
./appdata_ghijkl/js/core/merged-login.js.deps
./appdata_ghijkl/js/core/merged-login.js.gzip
./appdata_ghijkl/js/core/merged-share-backend.js
./appdata_ghijkl/js/core/merged-share-backend.js.deps
./appdata_ghijkl/js/core/merged-share-backend.js.gzip
./appdata_ghijkl/js/core/merged-template-prepend.js
./appdata_ghijkl/js/core/merged-template-prepend.js.deps
./appdata_ghijkl/js/core/merged-template-prepend.js.gzip
./appdata_ghijkl/js/files
./appdata_ghijkl/js/files/merged-index.js
./appdata_ghijkl/js/files/merged-index.js.deps
./appdata_ghijkl/js/files/merged-index.js.gzip
./appdata_ghijkl/js/files_sharing
./appdata_ghijkl/js/files_sharing/additionalScripts.js
./appdata_ghijkl/js/files_sharing/additionalScripts.js.deps
./appdata_ghijkl/js/files_sharing/additionalScripts.js.gzip
./appdata_ghijkl/js/files_texteditor
./appdata_ghijkl/js/files_texteditor/merged.js
./appdata_ghijkl/js/files_texteditor/merged.js.deps
./appdata_ghijkl/js/files_texteditor/merged.js.gzip
./appdata_ghijkl/js/files_versions
./appdata_ghijkl/js/files_versions/merged.js
./appdata_ghijkl/js/files_versions/merged.js.deps
./appdata_ghijkl/js/files_versions/merged.js.gzip
./appdata_ghijkl/js/gallery
./appdata_ghijkl/js/gallery/scripts-for-file-app.js
./appdata_ghijkl/js/gallery/scripts-for-file-app.js.deps
./appdata_ghijkl/js/gallery/scripts-for-file-app.js.gzip
./appdata_ghijkl/js/notifications
./appdata_ghijkl/js/notifications/merged.js
./appdata_ghijkl/js/notifications/merged.js.deps
./appdata_ghijkl/js/notifications/merged.js.gzip
./appdata_ghijkl/js/systemtags
./appdata_ghijkl/js/systemtags/merged.js
./appdata_ghijkl/js/systemtags/merged.js.deps
./appdata_ghijkl/js/systemtags/merged.js.gzip
./appdata_ghijkl/preview
./appdata_ghijkl/preview/7
./appdata_ghijkl/preview/7/2048-2048-max.png
./appdata_ghijkl/preview/7/32-32-crop.png
./appdata_ghijkl/theming
./appdata_ghijkl/theming/0
./appdata_ghijkl/theming/0/icon-core-filetypes_application-pdf.svg
./appdata_ghijkl/theming/0/icon-core-filetypes_file.svg
./appdata_ghijkl/theming/0/icon-core-filetypes_folder.svg
./appdata_ghijkl/theming/0/icon-core-filetypes_text.svg
./appdata_ghijkl/theming/0/icon-core-filetypes_video.svg
./appdata_ghijkl/theming/0/icon-core-filetypes_x-office-document.svg
./files_external
./files_external/rootcerts.crt
./index.html
./nextcloud.log

238 directories, 75 files
```

</details>

cc @rullzer @icewind1991 @juliushaertl @oparoz @karlitschek

**Comment by MorrisJobke** [comment]
![abc](https://cloud.githubusercontent.com/assets/245432/25765081/78727172-31c1-11e7-833b-d1c189fcf3d7.png)


Left is node1 and right is node2. On node1 a new file was added and the preview is only available there.

**Comment by karlitschek** [comment]
Caches 'might' be O.K. to be local. Might be faster in some scenarios. But content that only exist once like avatars have to be in the object store of course.

**Comment by rullzer** [comment]
@icewind1991 you probabaly have a better idea how much work it would be to have appdata on objectstore as well...?

**Comment by MorrisJobke** [comment]
> Caches 'might' be O.K. to be local. Might be faster in some scenarios. But content that only exist once like avatars have to be in the object store of course.

The problem is not, that they are recreated, but that we store info about this in the file cache and this is shared. So node1 creates an preview and adds it to the file cache. On node2 it is not available and 💥 (either it cleans up the file cache and creates a new local file or it returns a 404 because the file cache says "yes" but the filesystem "no" - I don't know how it behaves, but in both ways it's not good)

**Comment by rullzer** [comment]
Probably it does boom...
Also in case of previews they need to be shared. Since there is no way to invalidate them on other hosts

**Comment by icewind1991** [comment]
This should already be the case.

On a new instance with s3 configured, after uploading some files, creating a user, setting avatar. I have the following the datadir

```
total 648
-rw-r--r-- 1 http http      0 mei  8 14:12 index.html
-rw-r----- 1 http http    732 mei  8 14:13 nextcloud.log
-rw-r--r-- 1 http http 659456 mei  8 14:13 owncloud.db
```

**Comment by icewind1991** [comment]
@MorrisJobke did you setup the objectstore before installing or after?

**Comment by MorrisJobke** [comment]
Yep. I used an S3 storage. :/ the screenshots from above are from this instance.

**Comment by MorrisJobke** [comment]
Ah - the yes should mean: we placed the object storage config in the config.php and then ran the install command.

**Comment by MorrisJobke** [comment]
The code was the daily tarball of master from Friday

**Comment by rullzer** [comment]
@MorrisJobke this is fixed by #4798 right?

**Comment by MorrisJobke** [comment]
Yep
